cmdlib: fallback to virt when on overlayfs

jlebon opened the pull request:
Really, we can support overlayfs fine if we just do the compose in virt.
This is important in CI contexts where code runs in a pod and a PVC
mount might not be available.

(Of course, in *most* CI contexts, we're already not privileged so we
weren't hitting this.)

A more generic approach here would just be to test xattr capabilities
given the target filesystem and environment we're in.

cgwalters commented:
Hmm.  You don't need a PVC, can just use an [emptyDir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir).  Why wouldn't you do that?

jlebon commented:
> Hmm. You don't need a PVC, can just use an emptyDir. Why wouldn't you do that?

I just didn't think of it. ;)

The underlying context here though is that I want to be able to `cosa fetch --update-lockfile --dry-run` (https://github.com/coreos/coreos-assembler/pull/828) and so I don't actually need to compose at all (in fact, we don't even need supermin/sudo for this; I have a follow-up patch to just be able to run `rpm-ostree` directly). I added this context to the commit message.

I guess I could request an `emptyDir`, but that seems unnecessarily harder than it needs to be if I just want to update the lockfile. An alternative is to just drop the check altogether and let rpm-ostree error out (since we're erroring out anyway) or to put it in `cmd-build`.

> The reason I added that check wasn't just because things didn't work on overlayfs - it's because I wanted to discourage writing builds into transient container storage, and have people be surprised when their builds vanish when the container is destroyed.

But building into transient container storage is a totally fine thing to do in a CI context, right? How is that different from using an `emptyDir` mount which gets auto-destroyed when the pod exits?

cgwalters commented:
> But building into transient container storage is a totally fine thing to do in a CI context, right? How is that different from using an emptyDir mount which gets auto-destroyed when the pod exits?

First, talking to a *real* underlying filesystem directly is just better.  The real kernel overlayfs has had bugs in the past, and so has [fuse-overlayfs](https://github.com/containers/fuse-overlayfs/issues/104).  

The data we're writing has nothing to do with the container image, so why should it be on the same filesystem.

But again the main thing for me here is to avoid having people miss a `-v` argument to `podman` ending up losing their builds.

> I guess I could request an emptyDir, but that seems unnecessarily harder than it needs to be

I don't see what's hard about using `emptyDir`.

cgwalters commented:
I guess one thing we could do is move this check down into places where we actually *write* data?  So `--dry-run` wouldn't require an transient mount.

jlebon commented:
> But again the main thing for me here is to avoid having people miss a -v argument to podman ending up losing their builds.

>> I guess I could request an emptyDir, but that seems unnecessarily harder than it needs to be
>
> I don't see what's hard about using emptyDir.

Sorry, I'm trying to follow. What I'm saying is that using `emptyDir` has the exact same issue that not issuing `-v` has wrt transience. This also assumes that users are using `--rm`. (And if they're using the `cosa` alias, then it already has all the right flags in place).

But anyway, I don't actually *need* to compose into overlayfs, I just need it to be outside the `cosa fetch --dry-run` code path. I don't disagree that it's not the best way to run the compose, though at the same time I find it odd to go out of our way to forbid it.

> I guess one thing we could do is move this check down into places where we actually write data?

Yup, I suggested higher up making it part of `cmd-build`. Does that work for you?

cgwalters commented:
> Yup, I suggested higher up making it part of cmd-build. Does that work for you?

And all `buildextend-` right?

jlebon commented:
I don't think this is worth it; will just work around it.
